<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>What is CrossMap ? &mdash; CrossMap documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=dc072ff8"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: white" >

          
          
          <a href="#" class="icon icon-home">
            CrossMap
              <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">What is CrossMap ?</a></li>
<li><a class="reference internal" href="#how-crossmap-works">How CrossMap works?</a></li>
<li><a class="reference internal" href="#release-history">Release history</a></li>
<li><a class="reference internal" href="#installation">Installation</a><ul>
<li><a class="reference internal" href="#install-from-pypi">Install from PyPI</a></li>
<li><a class="reference internal" href="#install-from-github">Install from GitHub</a></li>
<li><a class="reference internal" href="#upgrade">Upgrade</a></li>
</ul>
</li>
<li><a class="reference internal" href="#input-and-output">Input and Output</a><ul>
<li><a class="reference internal" href="#chain-file">Chain file</a></li>
<li><a class="reference internal" href="#user-input-file">User Input file</a></li>
<li><a class="reference internal" href="#output-file">Output file</a></li>
</ul>
</li>
<li><a class="reference internal" href="#usage">Usage</a><ul>
<li><a class="reference internal" href="#convert-bed-format-files">Convert BED format files</a></li>
<li><a class="reference internal" href="#convert-bam-cram-sam-format-files">Convert BAM/CRAM/SAM format files</a></li>
<li><a class="reference internal" href="#convert-wiggle-format-files">Convert Wiggle format files</a></li>
<li><a class="reference internal" href="#convert-bigwig-format-files">Convert BigWig format files</a></li>
<li><a class="reference internal" href="#convert-gff-gtf-format-files">Convert GFF/GTF format files</a></li>
<li><a class="reference internal" href="#convert-vcf-format-files">Convert VCF format files</a></li>
<li><a class="reference internal" href="#convert-maf-format-files">Convert MAF format files</a></li>
<li><a class="reference internal" href="#convert-gvcf-format-files">Convert GVCF format files</a></li>
<li><a class="reference internal" href="#convert-large-genomic-regions">Convert large genomic regions</a></li>
<li><a class="reference internal" href="#view-chain-file">View chain file</a></li>
</ul>
</li>
<li><a class="reference internal" href="#compare-to-ucsc-liftover-tool">Compare to UCSC liftover tool</a></li>
<li><a class="reference internal" href="#citation">Citation</a></li>
<li><a class="reference internal" href="#license">LICENSE</a></li>
<li><a class="reference internal" href="#contact">Contact</a></li>
</ul>
</div>
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: white" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">CrossMap</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">What is CrossMap ?</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="toctree-wrapper compound">
</div>
<section id="what-is-crossmap">
<h1>What is CrossMap ?<a class="headerlink" href="#what-is-crossmap" title="Permalink to this heading"></a></h1>
<p>CrossMap is a program for genome coordinates conversion between <em>different assemblies</em>
(such as <a class="reference external" href="http://www.ncbi.nlm.nih.gov/assembly/2928/">hg18 (NCBI36)</a>   &lt;=&gt; <a class="reference external" href="http://www.ncbi.nlm.nih.gov/assembly/2758/">hg19 (GRCh37)</a>).
It supports commonly used file formats including <a class="reference external" href="https://samtools.github.io/hts-specs/SAMv1.pdf">BAM</a>, <a class="reference external" href="https://en.wikipedia.org/wiki/CRAM_(file_format)">CRAM</a>, <a class="reference external" href="https://en.wikipedia.org/wiki/SAM_(file_format)">SAM</a>, <a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/wiggle.html">Wiggle</a>, <a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/bigWig.html">BigWig</a>, <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED</a>, <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format3">GFF</a>, <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format4">GTF</a>, <a class="reference external" href="https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/">MAF</a> <a class="reference external" href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">VCF</a>, and <a class="reference external" href="https://sites.google.com/site/gvcftools/home/about-gvcf">gVCF</a>.</p>
</section>
<section id="how-crossmap-works">
<h1>How CrossMap works?<a class="headerlink" href="#how-crossmap-works" title="Permalink to this heading"></a></h1>
<a class="reference internal image-reference" href="_images/howitworks.png"><img alt="alternate text" src="_images/howitworks.png" style="width: 510.0px; height: 212.5px;" /></a>
</section>
<section id="release-history">
<h1>Release history<a class="headerlink" href="#release-history" title="Permalink to this heading"></a></h1>
<p><strong>07/17/2024: Release version 0.7.3</strong></p>
<ol class="arabic simple">
<li><p>Fix bugs for VCF (and gVCF) liftover. For variants with multiple ALT alleles, remove the ALT allele that is the same as the REF allele.</p></li>
</ol>
<p><strong>05/09/2024: Release version 0.7.2</strong></p>
<ol class="arabic simple">
<li><p>Fix bugs for VCF (and gVCF) liftover. When insertion/deletion variants were mapped to the reverse region of the target assembly, their REF alleles need to update.</p></li>
</ol>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span># Input VCF (header was not shown)
# These are hg19/GRCh37 based varants. They will map to the reverse region on hg38/GRCh38
chr7    61879851        rs1223781306    A       AC      .       .       .
chr1    145382743       rs782203468     G       GA      .       .       .
chr1    144852392       indel.6062      AC      A       .       .       .
chr1    145698920       indel.6189      TGCTTGGGGTGCTTACG       T       .       .       .

# Output (Note their REF alleles were different from input)
chr7    62217710        rs1223781306    T       TG      .       .       .
chr1    146052257       rs782203468     C       CT      .       .       .
chr1    149032049       indel.6062      GG      G       .       .       .
chr1    145736150       indel.6189      CCGTAAGCACCCCAAGC       C       .       .       .
</pre></div>
</div>
<p><strong>01/11/2024: Release version 0.7.0</strong></p>
<ol class="arabic simple">
<li><p>Fix bugs for VCF varaints liftover.</p></li>
<li><p>Handle non-DNA ALT alleles such as &lt;DEL&gt;</p></li>
<li><p>Use <a class="reference external" href="https://packaging.python.org/en/latest/guides/writing-pyproject-toml/">pyproject.toml</a> to replace “setup.py”.</p></li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>From v0.7.0 onwards, the main program <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span></code> is now renamed to <code class="code docutils literal notranslate"><span class="pre">CrossMap</span></code>
due to the restriction on using “.” in the script name in the “pyproject.toml” file.
To ensure compatibility with the previous pipelines, please include the following line in
your <code class="docutils literal notranslate"><span class="pre">~/.bashrc</span></code> file.</p>
</div>
<p><code class="docutils literal notranslate"><span class="pre">alias</span> <span class="pre">CrossMap.py='CrossMap'</span></code></p>
<p><strong>07/21/2023: Release version 0.6.4</strong></p>
<ol class="arabic simple">
<li><p>Fix bug when the sequence in BAM file is represented as “*”</p></li>
<li><p>Change code style</p></li>
</ol>
<p><strong>07/12/2022: Release version 0.6.4</strong></p>
<ol class="arabic simple">
<li><p>Fix bug when the input bigwig file does not have coverage signal for some chromosomes.</p></li>
<li><p>When the input VCF file does not have CONTIG field, use long chromosome ID (e.g., “chr1”) as default.</p></li>
</ol>
<p><strong>03/04/2022: Release version 0.6.3</strong></p>
<ol class="arabic simple">
<li><p>Fix bug in v0.6.2. “Alternative allele is empty”</p></li>
</ol>
<p><strong>02/22/2022: Release version 0.6.2</strong></p>
<ol class="arabic simple">
<li><p>For insertions and deletions, the first nucleotide of the ALT allele (the 5th field in VCF file) is updated to the nucleotide at POS of the reference genome</p></li>
</ol>
<p><strong>11/29/2021: Release version 0.6.1</strong></p>
<ol class="arabic simple">
<li><p>Same as v0.6.0. Remove unused modules from the lib folder.</p></li>
</ol>
<p><strong>11/16/2021: Release version 0.6.0</strong></p>
<ol class="arabic simple">
<li><p>Use <a class="reference external" href="https://docs.python.org/3/library/argparse.html">argparse</a> instead of <a class="reference external" href="https://docs.python.org/3/library/optparse.html">optparse</a>.</p></li>
<li><p>Use <code class="code docutils literal notranslate"><span class="pre">os.path.getmtime</span></code> instead of <code class="code docutils literal notranslate"><span class="pre">os.path.getctime</span></code> to check the timestamps of fasta file and its index file.</p></li>
<li><p>Add ‘–unmap-file’ option to <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">bed</span></code> command.</p></li>
</ol>
<p><strong>4/16/2021: Release version 0.5.3/0.5.4</strong></p>
<p>Add <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">viewchain</span></code> to convert chain file into block-to-block, more readable format.</p>
<p><strong>12/08/2020: Release version 0.5.2</strong></p>
<p>Add ‘–no-comp-alleles’ flag to <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">vcf</span></code> and <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">gvcf</span></code>. If set, CrossMap does not check if the “reference allele” is different from the “alternative allele”.</p>
<p><strong>08/19/2020: Release version 0.5.1</strong></p>
<p>In <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">region</span></code>: keep additional columns (columns after the 3rd column) of the original BED file after conversion.</p>
<p><strong>08/14/2020: Release version 0.5.0</strong></p>
<p>Add <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">region</span></code> function to convert large genomic regions. Unlike the <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">bed</span></code> function, which splits big genomic regions, <code class="code docutils literal notranslate"><span class="pre">CrossMap.py</span> <span class="pre">region</span></code> tries to convert the big genomic region as a whole.</p>
<p><strong>07/09/2020: Release version 0.4.3</strong></p>
<p>Structural Variants VCF files often use INFO/END field to indicate the end of a deletion. v0.4.3 updates “END” coordinate in the INFO field.</p>
<p><strong>05/04/2020: Release version 0.4.2</strong></p>
<p>Support <a class="reference external" href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531812-GVCF-Genomic-Variant-Call-Format">GVCF</a> file conversion.</p>
<p><strong>03/24/2020: Release version 0.4.1</strong></p>
<p>Deal with consecutive TABs in the input MAF file.</p>
<p><strong>10/09/2019: Release version 0.3.8</strong></p>
<p>The University of California holds the copyrights in the UCSC chain files. As requested by UCSC, all UCSC-generated chain files will be permanently removed from this website and the CrossMap distributions.</p>
<p><strong>07/22/2019: Release version 0.3.6</strong></p>
<ol class="arabic simple">
<li><p>Support MAF (mutation annotation format).</p></li>
<li><p>Fix error “TypeError: AlignmentHeader does not support item assignment (use header.to_dict()” when lifting over BAM files. User does not need to downgrade pysam to 0.13.0 to lift over BAM files.</p></li>
</ol>
<p><strong>04/01/2019: Release version 0.3.4</strong></p>
<p>Fix bugs when chromosome IDs (of the source genome) in chain file do not have ‘chr’ prefix (such as “GRCh37ToHg19.over.chain.gz”). This version also allows CrossMap to detect if a VCF mapping was inverted, and if so, reverse complements the alternative allele (Thanks to Andrew Yates). Improve wording.</p>
<p><strong>01/07/2019: Release version 0.3.3</strong></p>
<p>Version 0.3.3 is exactly the same as Version 0.3.2. The reason to release this version is that CrossMap-0.3.2.tar.gz was broken when uploading to pypi.</p>
<p><strong>12/14/18: Release version 0.3.2</strong></p>
<p>Fix the key error problem (e.g  <em>KeyError: “sequence ‘b’7_KI270803v1_alt’’ not present”</em>). This error happens when a locus from the original assembly is mapped to an “alternative”, “unplaced” or “unlocalized” contig in the target assembly, and this “target contig” does not exist in your target_ref.fa. In version 0.3.2, such loci will be silently skipped and saved to the “.unmap” file.</p>
<p><strong>11/05/18: Release version 0.3.0</strong></p>
<ol class="arabic simple">
<li><p>v0.3.0 or newer will Support Python3. Previous versions support Python2.7.*</p></li>
<li><p>add <a class="reference external" href="https://github.com/deeptools/pyBigWig">pyBigWig</a> as a dependency.</p></li>
</ol>
</section>
<section id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this heading"></a></h1>
<section id="install-from-pypi">
<h2>Install from PyPI<a class="headerlink" href="#install-from-pypi" title="Permalink to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">pip3</span> <span class="pre">install</span> <span class="pre">CrossMap</span></code></p>
</section>
<section id="install-from-github">
<h2>Install from GitHub<a class="headerlink" href="#install-from-github" title="Permalink to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">pip3</span> <span class="pre">install</span> <span class="pre">git+https://github.com/liguowang/CrossMap.git</span></code></p>
</section>
<section id="upgrade">
<h2>Upgrade<a class="headerlink" href="#upgrade" title="Permalink to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">pip3</span> <span class="pre">install</span> <span class="pre">CrossMap</span> <span class="pre">--upgrade</span></code></p>
</section>
</section>
<section id="input-and-output">
<h1>Input and Output<a class="headerlink" href="#input-and-output" title="Permalink to this heading"></a></h1>
<section id="chain-file">
<h2>Chain file<a class="headerlink" href="#chain-file" title="Permalink to this heading"></a></h2>
<p>A <a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/chain.html">chain file</a> describes a pairwise alignment between two reference assemblies. <a class="reference external" href="https://genome.ucsc.edu/">UCSC</a> and <a class="reference external" href="https://uswest.ensembl.org/index.html">Ensembl</a> chain files are available:</p>
<p><strong>UCSC chain files</strong></p>
<blockquote>
<div><ul class="simple">
<li><p>Chain files from hs1 (T2T-CHM13) to hg38/hg19/mm10/mm9 (ore vice versa): <a class="reference external" href="https://hgdownload.soe.ucsc.edu/goldenPath/hs1/liftOver/">https://hgdownload.soe.ucsc.edu/goldenPath/hs1/liftOver/</a></p></li>
<li><p>Chain files from hg38 (GRCh38) to hg19 and all other organisms: <a class="reference external" href="http://hgdownload.soe.ucsc.edu/goldenPath/hg38/liftOver/">http://hgdownload.soe.ucsc.edu/goldenPath/hg38/liftOver/</a></p></li>
<li><p>Chain File from hg19 (GRCh37) to hg17/hg18/hg38 and all other organisms: <a class="reference external" href="http://hgdownload.soe.ucsc.edu/goldenPath/hg19/liftOver/">http://hgdownload.soe.ucsc.edu/goldenPath/hg19/liftOver/</a></p></li>
<li><p>Chain File from mm10 (GRCm38) to mm9 and all other organisms: <a class="reference external" href="http://hgdownload.soe.ucsc.edu/goldenPath/mm10/liftOver/">http://hgdownload.soe.ucsc.edu/goldenPath/mm10/liftOver/</a></p></li>
</ul>
</div></blockquote>
<p><strong>Ensembl chain files</strong></p>
<blockquote>
<div><ul class="simple">
<li><p>Human to Human: <a class="reference external" href="ftp://ftp.ensembl.org/pub/assembly_mapping/homo_sapiens/">ftp://ftp.ensembl.org/pub/assembly_mapping/homo_sapiens/</a></p></li>
<li><p>Mouse to Mouse: <a class="reference external" href="ftp://ftp.ensembl.org/pub/assembly_mapping/mus_musculus/">ftp://ftp.ensembl.org/pub/assembly_mapping/mus_musculus/</a></p></li>
<li><p>Other organisms: <a class="reference external" href="ftp://ftp.ensembl.org/pub/assembly_mapping/">ftp://ftp.ensembl.org/pub/assembly_mapping/</a></p></li>
</ul>
</div></blockquote>
</section>
<section id="user-input-file">
<h2>User Input file<a class="headerlink" href="#user-input-file" title="Permalink to this heading"></a></h2>
<p>CrossMap supports the following file formats.</p>
<ol class="arabic simple">
<li><p><a class="reference external" href="http://samtools.sourceforge.net/SAMv1.pdf">BAM</a>, <a class="reference external" href="https://samtools.github.io/hts-specs/CRAMv3.pdf">CRAM</a>, or <a class="reference external" href="http://samtools.sourceforge.net/SAMv1.pdf/">SAM</a></p></li>
<li><p><a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED</a> or BED-like. (BED file must have at least ‘chrom’, ‘start’, ‘end’)</p></li>
<li><p><a class="reference external" href="http://genome.ucsc.edu/goldenPath/help/wiggle.html">Wiggle</a> (“variableStep”, “fixedStep” and “bedGraph” formats are supported)</p></li>
<li><p><a class="reference external" href="http://genome.ucsc.edu/goldenPath/help/bigWig.html">BigWig</a></p></li>
<li><p><a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format3">GFF</a> or <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format4">GTF</a></p></li>
<li><p><a class="reference external" href="http://vcftools.sourceforge.net/index.html">VCF</a></p></li>
<li><p><a class="reference external" href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531812-GVCF-Genomic-Variant-Call-Format">GVCF</a></p></li>
<li><p><a class="reference external" href="https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/">MAF</a></p></li>
</ol>
</section>
<section id="output-file">
<h2>Output file<a class="headerlink" href="#output-file" title="Permalink to this heading"></a></h2>
<p>The format of output files depends on the input</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Input_format</p></th>
<th class="head"><p>Output_format</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>BED</p></td>
<td><p>BED (Genome coordinates will be updated)</p></td>
</tr>
<tr class="row-odd"><td><p>BAM</p></td>
<td><p>BAM (Genome coordinates, header section, all SAM flags, insert size will be updated)</p></td>
</tr>
<tr class="row-even"><td><p>CRAM</p></td>
<td><p>BAM (require pysam &gt;= 0.8.2)</p></td>
</tr>
<tr class="row-odd"><td><p>SAM</p></td>
<td><p>SAM (Genome coordinates, header section, all SAM flags, insert size will be updated)</p></td>
</tr>
<tr class="row-even"><td><p>Wiggle</p></td>
<td><p>BigWig</p></td>
</tr>
<tr class="row-odd"><td><p>BigWig</p></td>
<td><p>BigWig</p></td>
</tr>
<tr class="row-even"><td><p>GFF</p></td>
<td><p>GFF (Genome coordinates will be updated to the target assembly)</p></td>
</tr>
<tr class="row-odd"><td><p>GTF</p></td>
<td><p>GTF (Genome coordinates will be updated to the target assembly)</p></td>
</tr>
<tr class="row-even"><td><p>VCF</p></td>
<td><p>VCF (header section, Genome coordinates, reference alleles will be updated)</p></td>
</tr>
<tr class="row-odd"><td><p>GVCF</p></td>
<td><p>GVCF (header section, Genome coordinates, reference alleles will be updated)</p></td>
</tr>
<tr class="row-even"><td><p>MAF</p></td>
<td><p>MAF (Genome coordinates and reference alleles will be updated)</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this heading"></a></h1>
<p>Run <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">-h</span></code> or <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">--help</span></code> print help message</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap -h

usage: CrossMap [-h] [-v]
                {bed,bam,gff,wig,bigwig,vcf,gvcf,maf,region,viewchain} ...

CrossMap (v0.7.0) is a program to convert (liftover) genome coordinates
between different reference assemblies (e.g., from human GRCh37/hg19 to
GRCh38/hg38 or vice versa). Supported file formats: BAM, BED, BigWig, CRAM,
GFF, GTF, GVCF, MAF (mutation annotation format), SAM, Wiggle, and VCF.

positional arguments:
  {bed,bam,gff,wig,bigwig,vcf,gvcf,maf,region,viewchain}
                        sub-command help
    bed                 converts BED, bedGraph or other BED-like files. Only
                        genome coordinates (i.e., the first 3 columns) will be
                        updated. Regions mapped to multiple locations to the
                        new assembly will be split. Use the &quot;region&quot; command
                        to liftover large genomic regions. Use the &quot;wig&quot;
                        command if you need bedGraph/bigWig output.
    bam                 converts BAM, CRAM, or SAM format file. Genome
                        coordinates, header section, all SAM flags, insert
                        size will be updated.
    gff                 converts GFF or GTF format file. Genome coordinates
                        will be updated.
    wig                 converts Wiggle or bedGraph format file. Genome
                        coordinates will be updated.
    bigwig              converts BigWig file. Genome coordinates will be
                        updated.
    vcf                 converts VCF file. Genome coordinates, header section,
                        reference alleles will be updated.
    gvcf                converts GVCF file. Genome coordinates, header
                        section, reference alleles will be updated.
    maf                 converts MAF (mutation annotation format) file. Genome
                        coordinates and reference alleles will be updated.
    region              converts big genomic regions (in BED format) such as
                        CNV blocks. Genome coordinates will be updated.
    viewchain           prints out the content of a chain file into a human
                        readable, block-to-block format.

options:
  -h, --help            show this help message and exit
  -v, --version         show program&#39;s version number and exit

https://crossmap.readthedocs.io/en/latest/
</pre></div>
</div>
<section id="convert-bed-format-files">
<h2>Convert BED format files<a class="headerlink" href="#convert-bed-format-files" title="Permalink to this heading"></a></h2>
<p>A <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED</a> (Browser Extensible Data) file
is a tab-delimited text file describing genome regions or gene annotations.
It consists of one line per feature, each containing 3-12 columns.
CrossMap converts BED files with less than 12 columns to a different assembly by updating the
chromosome and genome coordinates only; all other columns remain unchanged. Regions from the old
assembly mapping to multiple locations to the new assembly will be split.  For 12-columns BED
files, all columns will be updated accordingly except the 4th column (name of bed line), 5th
column (score value), and 9th column (RGB value describing the display color). 12-column BED
files usually define multiple blocks (e.g., exons); if any of the exons fails to map to a new
assembly, the whole BED line is skipped.</p>
<p>The input BED file can be plain text file, compressed file with extension of .gz, .Z, .z,
.bz, .bz2 and .bzip2, or even a URL pointing to accessible remote files (<a class="reference external" href="http://">http://</a>, <a class="reference external" href="https://">https://</a>
and <a class="reference external" href="ftp://">ftp://</a>). Compressed remote files are not supported. The output is a BED format file with
exactly the same number of columns as the original one.</p>
<p>Standard <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED</a> format has 12 columns, but CrossMap also supports BED-like formats:</p>
<ul class="simple">
<li><p>BED3: The first three columns (“chrom”, “start”, “end”) of the BED format file.</p></li>
<li><p>BED6: The first six columns (“chrom”, “start”, “end”, “name”, “score”, “strand”) of the BED format file.</p></li>
<li><p>Other: Format has at least three columns (“chrom”, “start”, “end”) and no more than 12 columns. All other columns are arbitrary.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ol class="arabic simple">
<li><p>For BED-like formats mentioned above, CrossMap only updates the “chrom”, “start”, “end”, and “strand” columns. All other columns will be kept AS-IS.</p></li>
<li><p>Lines starting with ‘#’, ‘browser’, ‘track’ will be skipped.</p></li>
<li><p>Lines less than three columns will be skipped.</p></li>
<li><p>The 2nd and 3rd columns must be integers.</p></li>
<li><p>The “+” strand is assumed if no strand information is found.</p></li>
<li><p>For standard BED format (12 columns). If any of the defined exon blocks cannot be uniquely mapped to target assembly, the whole entry will be skipped.</p></li>
<li><p>The “input_chain_file” and “input_bed_file” can be regular or compressed (.gz, .Z, .z, .bz, .bz2, .bzip2) file, local file or URL (<a class="reference external" href="http://">http://</a>, <a class="reference external" href="https://">https://</a>, <a class="reference external" href="ftp://">ftp://</a>) pointing to remote file.</p></li>
<li><p>If the output_file is not specified, results will be printed to screen (console). In this case, the original bed entries (including entries failed to convert) were also printed out.</p></li>
<li><p>If the input region cannot be consecutively mapped to the target assembly, it will be split.</p></li>
<li><p>The *.unmap file contains regions that cannot be unambiguously converted.</p></li>
</ol>
</div>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">bed</span> <span class="pre">-h</span></code> will print help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap bed  -h
usage: CrossMap bed [-h] [--chromid {a,s,l}] [--unmap-file UNMAP_FILE]
                    input.chain input.bed [out_bed]

positional arguments:
  input.chain           Chain file
                        (https://genome.ucsc.edu/goldenPath/help/chain.html)
                        describes pairwise alignments between two genomes. The
                        input chain file can be a plain text file or
                        compressed (.gz, .Z, .z, .bz, .bz2, .bzip2) file.
  input.bed             The input BED file. The first 3 columns must be
                        “chrom”, “start”, and “end”. The input BED file can be
                        plain text file, compressed file with extension of
                        .gz, .Z, .z, .bz, .bz2 and .bzip2, or even a URL
                        pointing to accessible remote files (http://, https://
                        and ftp://). Compressed remote files are not
                        supported.
  out_bed               Output BED file. if argument is missing, CrossMap will
                        write BED file to the STDOUT.

options:
  -h, --help            show this help message and exit
  --chromid {a,s,l}     The style of chromosome IDs. &quot;a&quot; = &quot;as-is&quot;; &quot;l&quot; =
                        &quot;long style&quot; (eg. &quot;chr1&quot;, &quot;chrX&quot;); &quot;s&quot; = &quot;short style&quot;
                        (eg. &quot;1&quot;, &quot;X&quot;).
  --unmap-file UNMAP_FILE
                        file to save unmapped entries. This will be ignored if
                        [out_bed] was not provided.
</pre></div>
</div>
<p><strong>Example 1</strong></p>
<p>run <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">bed</span></code> with no output file.
Results were printed to screen.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$cat Test1.hg19.bed

chr1   65886334    66103176
chr3    112251353   112280810
chr5    54408798    54469005
chr7    107204401   107218968

$ CrossMap bed GRCh37_to_GRCh38.chain.gz Test1.hg19.bed

2024-01-12 08:36:35 [INFO]  Read the chain file &quot;GRCh37_to_GRCh38.chain.gz&quot;
chr1   65886334    66103176    -&gt;  chr1    65420651    65637493
chr3   112251353   112280810   -&gt;  chr3    112532506   112561963
chr5   54408798    54469005    -&gt;  chr5    55112970    55173177
chr7   107204401   107218968   -&gt;  chr7    107563956   107578523
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The first 3 columns are hg19-based coordinates. The last 3 columns are hg38-based coordinates.</p>
</div>
<p><strong>Example 2</strong></p>
<p>run <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">bed</span></code> with output file specified.
Results (genomic coordinates after liftover) were saved to the file “output.hg38”</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ CrossMap bed GRCh37_to_GRCh38.chain.gz Test1.hg19.bed output.hg38
2024-01-12 08:35:56 [INFO]  Read the chain file &quot;GRCh37_to_GRCh38.chain.gz&quot;

$ cat output.hg38
chr1    65420651    65637493
chr3    112532506   112561963
chr5    55112970    55173177
chr7    107563956   107578523
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Genomic intervals failed to map will be saved to file “output.hg38.unmap”.</p>
</div>
<p><strong>Example 3</strong></p>
<p>Input regions will be split if they cannot map consecutively to the target assembly.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ cat Test2.hg19.bed
chr20  21106623    21227258
chr22  30792929    30821291

$ CrossMap bed GRCh37_to_GRCh38.chain.gz Test2.hg19.bed
2024-01-12 08:53:10 [INFO]  Read the chain file &quot;GRCh37_to_GRCh38.chain.gz&quot;
chr20  21106623    21227258    (split.1:chr20:21106623:21144549:+) chr20   21125982    21163908
chr20  21106623    21227258    (split.2:chr20:21144549:21176014:+) chr20   21163909    21195374
chr20  21106623    21227258    (split.3:chr20:21176014:21186161:+) chr20   21195375    21205522
chr20  21106623    21227258    (split.4:chr20:21186161:21227258:+) chr20   21205523    21246620
chr22  30792929    30821291    (split.1:chr22:30792929:30819612:+) chr22   30396940    30423623
chr22  30792929    30821291    (split.2:chr22:30819615:30821291:+) chr22   30423627    30425303
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The first 3 columns are hg19-based coordinates. The last 3 columns are hg38-based coordinates.</p>
</div>
<p><strong>Example 4</strong></p>
<p><a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/bedgraph.html">BedGraph</a> format file can be converted using either <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">bed</span></code> or <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">wig</span></code>;
however, the output formats are different:</p>
<ul class="simple">
<li><p>When using <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">bed</span></code> command to convert a bedGraph file, the output is a <strong>bedGraph</strong> file.</p></li>
<li><p>When using <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">wig</span></code> command to convert a bedGraph file, the output is a <strong>bigWig</strong> file.</p></li>
</ul>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ head -3 Test3.hg19.bgr
chrX   2705083 2705158 1.0
chrX    2813094 2813169 0.9
chrX    2813169 2814363 0.1
...

$ CrossMap bed hg19ToHg38.over.chain.gz Test3.hg19.bgr
2024-01-12 09:05:05 [INFO]  Read the chain file &quot;GRCh37_to_GRCh38.chain.gz
chrX    2705083 2705158 1.0 -&gt;  chrX    2787042 2787117 1.0
chrX    2813094 2813169 0.9 -&gt;  chrX    2895053 2895128 0.9
chrX    2813169 2814363 0.1 -&gt;  chrX    2895128 2896322 0.1
...

$ CrossMap wig  GRCh37_to_GRCh38.chain.gz  Test3.hg19.bgr output.hg38
2024-01-12 09:09:52 [INFO]  Read the chain file &quot;GRCh37_to_GRCh38.chain.gz&quot;
2024-01-12 09:09:52 [INFO]  Liftover wiggle file &quot;Test3.hg19.bgr&quot; to bedGraph file &quot;output.hg38.bgr&quot;
2024-01-12 09:09:52 [INFO]  Merging overlapped entries in bedGraph file
2024-01-12 09:09:52 [INFO]  Sorting bedGraph file: output.hg38.bgr
2024-01-12 09:09:52 [INFO]  Writing header to &quot;output.hg38.bw&quot; ...
2024-01-12 09:09:52 [INFO]  Writing entries to &quot;output.hg38.bw&quot; ...
</pre></div>
</div>
<p><strong>Example 5</strong></p>
<p>Use <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">region</span></code> command to convert large genomic regions (such as <a class="reference external" href="https://en.wikipedia.org/wiki/Copy_number_variation">CNV</a> blocks) in BED format.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ cat Test4.hg19.bed
chr2   239716679       243199373  # A large genomic interval of 3.48 Mb
</pre></div>
</div>
<p>If use the :code:<a href="#id3"><span class="problematic" id="id4">`</span></a>CrossMap bed`command to liftover, this interval will be split 74 times.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ CrossMap bed GRCh37_to_GRCh38.chain.gz Test4.hg19.bed
2024-01-12 09:12:45 [INFO]  Read the chain file &quot;GRCh37_to_GRCh38.chain.gz&quot;
chr2   239716679       243199373       (split.1:chr2:239716679:239801978:+)    chr2    238808038       238893337
chr2   239716679       243199373       (split.2:chr2:239831978:240205681:+)    chr2    238910282       239283985
chr2   239716679       243199373       (split.3:chr2:240205681:240319336:+)    chr2    239283986       239397641
... (split 74 times)
</pre></div>
</div>
<p>Use the <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">region</span></code> command to liftover. By defualt <code class="code docutils literal notranslate"><span class="pre">r</span> <span class="pre">=</span> <span class="pre">0.85</span></code>.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ CrossMap region GRCh37_to_GRCh38.chain.gz Test4.hg19.bed
2024-01-12 09:15:24 [INFO]  Read the chain file &quot;GRCh37_to_GRCh38.chain.gz&quot;
chr2    239716679   243199373   -&gt;  chr2    238808038   242183529   map_ratio=0.9
</pre></div>
</div>
<p>If we set <code class="code docutils literal notranslate"><span class="pre">r</span> <span class="pre">=</span> <span class="pre">0.99</span></code>, this region will fail.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ CrossMap region GRCh37_to_GRCh38.chain.gz Test4.hg19.bed -r 0.99
2024-01-12 09:18:53 [INFO]  Read the chain file &quot;GRCh37_to_GRCh38.chain.gz&quot;
chr2    239716679   243199373   Fail    map_ratio=0.9622
</pre></div>
</div>
</section>
<section id="convert-bam-cram-sam-format-files">
<span id="bam-conversion"></span><h2>Convert BAM/CRAM/SAM format files<a class="headerlink" href="#convert-bam-cram-sam-format-files" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml#5">SAM</a> (Sequence Alignment Map) format
is a generic format for storing sequencing alignments, and BAM is the binary and compressed
version of SAM (<a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/25/16/2078.full">Li et al., 2009</a>).
<a class="reference external" href="https://samtools.github.io/hts-specs/CRAMv3.pdf">CRAM</a> was designed to be an efficient reference-based
alternative to the <a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml#5">SAM</a> and BAM file formats.
Most high-throughput sequencing  (HTS) alignments were in SAM/BAM format and many HTS analysis
tools work with SAM/BAM format. CrossMap updates chromosomes, genome coordinates, header
sections, and all SAM flags accordingly.  CrossMap’s version number is inserted into
the header section, along with the names of the original BAM file and the chain file.  For
pair-end sequencing, insert size is also recalculated. The input BAM file should be sorted
and indexed properly using Samtools (<a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/25/16/2078.full">Li et al., 2009</a>).
The output format is determined by the input format, and the BAM output will be sorted and indexed automatically.</p>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">bam</span> <span class="pre">-h</span></code> will print help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap bam -h

usage: CrossMap bam [-h] [-m INSERT_SIZE] [-s INSERT_SIZE_STDEV]
                    [-t INSERT_SIZE_FOLD] [-a] [--chromid {a,s,l}]
                    input.chain input.bam [out_bam]

positional arguments:
  input.chain           Chain file
                        (https://genome.ucsc.edu/goldenPath/help/chain.html)
                        describes pairwise alignments between two genomes. The
                        input chain file can be a plain text file or
                        compressed (.gz, .Z, .z, .bz, .bz2, .bzip2) file.
  input.bam             Input BAM file (https://genome.ucsc.edu/FAQ/FAQformat.
                        html#format5.1).
  out_bam               Output BAM file. if argument is missing, CrossMap will
                        write BAM file to the STDOUT.

options:
  -h, --help            show this help message and exit
  -m INSERT_SIZE, --mean INSERT_SIZE
                        Average insert size of pair-end sequencing (bp).
  -s INSERT_SIZE_STDEV, --stdev INSERT_SIZE_STDEV
                        Stanadard deviation of insert size.
  -t INSERT_SIZE_FOLD, --times INSERT_SIZE_FOLD
                        A mapped pair is considered as &quot;proper pair&quot; if both
                        ends mapped to different strand and the distance
                        between them is less then &#39;-t&#39; * stdev from the mean.
  -a, --append-tags     Add tag to each alignment in BAM file. Tags for pair-
                        end alignments include: QF = QC failed, NN = both
                        read1 and read2 unmapped, NU = read1 unmapped, read2
                        unique mapped, NM = read1 unmapped, multiple mapped,
                        UN = read1 uniquely mapped, read2 unmap, UU = both
                        read1 and read2 uniquely mapped, UM = read1 uniquely
                        mapped, read2 multiple mapped, MN = read1 multiple
                        mapped, read2 unmapped, MU = read1 multiple mapped,
                        read2 unique mapped, MM = both read1 and read2
                        multiple mapped. Tags for single-end alignments
                        include: QF = QC failed, SN = unmaped, SM = multiple
                        mapped, SU = uniquely mapped.
  --chromid {a,s,l}     The style of chromosome IDs. &quot;a&quot; = &quot;as-is&quot;; &quot;l&quot; =
                        &quot;long style&quot; (eg. &quot;chr1&quot;, &quot;chrX&quot;); &quot;s&quot; = &quot;short style&quot;
                        (eg. &quot;1&quot;, &quot;X&quot;).
</pre></div>
</div>
<p><strong>Example</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ CrossMap bam -a GRCh37_to_GRCh38.chain.gz Test5.hg19.bam output.hg38
Add tags: True
Insert size = 200.000000
Insert size stdev = 30.000000
Number of stdev from the mean = 3.000000
Add tags to each alignment = True
2024-01-12 09:29:11 [INFO]  Read the chain file &quot;GRCh37_to_GRCh38.chain.gz&quot;
[E::idx_find_and_load] Could not retrieve index file for &#39;Test5.hg19.bam&#39;
2024-01-12 09:29:11 [INFO]  Liftover BAM file &quot;Test5.hg19.bam&quot; to &quot;output.hg38.bam&quot;
2024-01-12 09:29:15 [INFO]  Done!
2024-01-12 09:29:15 [INFO]  Sort &quot;output.hg38.bam&quot; and save as &quot;output.hg38.sorted.bam&quot;
2024-01-12 09:29:15 [INFO]  Index &quot;output.hg38.sorted.bam&quot; ...

Total alignments:99914
    QC failed: 0
    Paired-end reads:
        R1 unique, R2 unique (UU): 96035
        R1 unique, R2 unmapp (UN): 3638
        R1 unique, R2 multiple (UM): 0
        R1 multiple, R2 multiple (MM): 0
        R1 multiple, R2 unique (MU): 230
        R1 multiple, R2 unmapped (MN): 11
        R1 unmap, R2 unmap (NN): 0
        R1 unmap, R2 unique (NU): 0
        R1 unmap, R2 multiple (NM): 0
</pre></div>
</div>
<p><strong>Optional tags:</strong></p>
<dl class="simple">
<dt>Q</dt><dd><p>QC. QC failed.</p>
</dd>
<dt>N</dt><dd><p>Unmapped. Originally unmapped or originally mapped but failed to lift over to new assembly.</p>
</dd>
<dt>M</dt><dd><p>Multiple mapped. Alignment can be lifted over to multiple places.</p>
</dd>
<dt>U</dt><dd><p>Unique mapped. Alignment can be lifted over to only 1 place.</p>
</dd>
</dl>
<p><strong>Tags for pair-end sequencing include:</strong></p>
<ul class="simple">
<li><p>QF = QC failed</p></li>
<li><p>NN = both read1 and read2 unmapped</p></li>
<li><p>NU = read1 unmapped, read2 unique mapped</p></li>
<li><p>NM = read1 unmapped, multiple mapped</p></li>
<li><p>UN = read1 uniquely mapped, read2 unmap</p></li>
<li><p>UU = both read1 and read2 uniquely mapped</p></li>
<li><p>UM = read1 uniquely mapped, read2 multiple mapped</p></li>
<li><p>MN = read1 multiple mapped, read2 unmapped</p></li>
<li><p>MU = read1 multiple mapped, read2 unique mapped</p></li>
<li><p>MM = both read1 and read2 multiple mapped</p></li>
</ul>
<p><strong>Tags for single-end sequencing include:</strong></p>
<ul class="simple">
<li><p>QF = QC failed</p></li>
<li><p>SN = unmaped</p></li>
<li><p>SM = multiple mapped</p></li>
<li><p>SU = uniquely mapped</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ol class="arabic simple">
<li><p>All alignments (mapped, partial mapped, unmapped, QC failed) will write to one file. Users can filter them by tags.</p></li>
<li><p>The header section will be updated to the target assembly.</p></li>
<li><p>Genome coordinates and all SAM flags in the alignment section will be updated to the target assembly.</p></li>
<li><p>If the input is a CRAM file, pysam version should &gt;= 0.8.2</p></li>
<li><p>Optional fields in the alignment section will not update.</p></li>
</ol>
</div>
</section>
<section id="convert-wiggle-format-files">
<h2>Convert Wiggle format files<a class="headerlink" href="#convert-wiggle-format-files" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="http://genome.ucsc.edu/goldenPath/help/wiggle.html">Wiggle</a> (WIG) format is useful in
displaying continuous data such as GC content and the reads intensity of high-throughput sequencing data.
BigWig is a self-indexed binary-format Wiggle file and has the advantage of supporting random access.
Input wiggle data can be in variableStep (for data with irregular intervals) or fixedStep
(for data with regular intervals). Regardless of the input, the output files are always in bedGraph
format.</p>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">wig</span> <span class="pre">-h</span></code> will print help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap  wig -h

usage: CrossMap wig [-h] [--chromid {a,s,l}] input.chain input.wig out_wig

positional arguments:
  input.chain        Chain file
                     (https://genome.ucsc.edu/goldenPath/help/chain.html)
                     describes pairwise alignments between two genomes. The
                     input chain file can be a plain text file or compressed
                     (.gz, .Z, .z, .bz, .bz2, .bzip2) file.
  input.wig          The input wiggle/bedGraph format file
                     (http://genome.ucsc.edu/goldenPath/help/wiggle.html).
                     Both &quot;variableStep&quot; and &quot;fixedStep&quot; wiggle lines are
                     supported. The input wiggle/bedGraph file can be plain
                     text file, compressed file with extension of .gz, .Z, .z,
                     .bz, .bz2 and .bzip2, or even a URL pointing to
                     accessible remote files (http://, https:// and ftp://).
                     Compressed remote files are not supported.
  out_wig            Output bedGraph file. Regardless of the input is wiggle
                     or bedGraph, the output file is always in bedGraph
                     format.

options:
  -h, --help         show this help message and exit
  --chromid {a,s,l}  The style of chromosome IDs. &quot;a&quot; = &quot;as-is&quot;; &quot;l&quot; = &quot;long
                     style&quot; (eg. &quot;chr1&quot;, &quot;chrX&quot;); &quot;s&quot; = &quot;short style&quot; (eg.
                     &quot;1&quot;, &quot;X&quot;).
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To improve performance, this script calls <a class="reference external" href="http://www.gnu.org/software/coreutils/manual/html_node/sort-invocation.html">GNU “sort”</a> command internally. If the “sort” command is not callable, CrossMap will exit.</p>
</div>
</section>
<section id="convert-bigwig-format-files">
<h2>Convert BigWig format files<a class="headerlink" href="#convert-bigwig-format-files" title="Permalink to this heading"></a></h2>
<p>If an input file is in BigWig format, the output is BigWig format if UCSC’s
‘<a class="reference external" href="http://hgdownload.cse.ucsc.edu/admin/exe/">wigToBigWig</a>’ executable can be found;
otherwise, the output file will be in bedGraph format.</p>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">bigwig</span> <span class="pre">-h</span></code> will print help message.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap bigwig -h

usage: CrossMap bigwig [-h] [--chromid {a,s,l}] input.chain input.bw output.bw

positional arguments:
  input.chain        Chain file
                     (https://genome.ucsc.edu/goldenPath/help/chain.html)
                     describes pairwise alignments between two genomes. The
                     input chain file can be a plain text file or compressed
                     (.gz, .Z, .z, .bz, .bz2, .bzip2) file.
  input.bw           The input bigWig format file
                     (https://genome.ucsc.edu/goldenPath/help/bigWig.html).
  output.bw          Output bigWig file.

options:
  -h, --help         show this help message and exit
  --chromid {a,s,l}  The style of chromosome IDs. &quot;a&quot; = &quot;as-is&quot;; &quot;l&quot; = &quot;long
                     style&quot; (eg. &quot;chr1&quot;, &quot;chrX&quot;); &quot;s&quot; = &quot;short style&quot; (eg.
                     &quot;1&quot;, &quot;X&quot;).
</pre></div>
</div>
<p>Example</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ CrossMap bigwig GRCh37_to_GRCh38.chain.gz Test6.hg19.bw output.hg38
2024-01-12 09:37:32 [INFO]  Read the chain file &quot;GRCh37_to_GRCh38.chain.gz&quot;
2024-01-12 09:37:33 [INFO]  Liftover bigwig file Test6.hg19.bw to bedGraph file output.hg38.bgr:
2024-01-12 09:37:33 [INFO]  Merging overlapped entries in bedGraph file
2024-01-12 09:37:33 [INFO]  Sorting bedGraph file: output.hg38.bgr
2024-01-12 09:37:33 [INFO]  Writing header to &quot;output.hg38.bw&quot; ...
2024-01-12 09:37:33 [INFO]  Writing entries to &quot;output.hg38.bw&quot; ...
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To improve performance, this script calls <a class="reference external" href="http://www.gnu.org/software/coreutils/manual/html_node/sort-invocation.html">GNU “sort”</a> command internally. If “sort” command does not exist, CrossMap will exit.</p>
</div>
</section>
<section id="convert-gff-gtf-format-files">
<h2>Convert GFF/GTF format files<a class="headerlink" href="#convert-gff-gtf-format-files" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format3">GFF</a> (General Feature Format) is another
plain text file used to describe gene structure. <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format4">GTF</a>
(Gene Transfer Format) is a refined version of GTF. The first eight fields are the same as
GFF. Plain text, compressed plain text, and URLs pointing to remote files are all supported.
Only chromosome and genome coordinates are updated. The format of the output is determined from
the input.</p>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">gff</span> <span class="pre">-h</span></code> will print help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap  gff -h
</pre></div>
</div>
<p>usage: CrossMap gff [-h] [–chromid {a,s,l}] input.chain input.gff [out_gff]</p>
<blockquote>
<div><dl>
<dt>positional arguments:</dt><dd><dl class="simple">
<dt>input.chain        Chain file</dt><dd><p>(<a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/chain.html">https://genome.ucsc.edu/goldenPath/help/chain.html</a>)
describes pairwise alignments between two genomes. The
input chain file can be a plain text file or compressed
(.gz, .Z, .z, .bz, .bz2, .bzip2) file.</p>
</dd>
<dt>input.gff          The input GFF (General Feature Format,</dt><dd><p><a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format3">http://genome.ucsc.edu/FAQ/FAQformat.html#format3</a>) or GTF
(Gene Transfer Format,
<a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format4">http://genome.ucsc.edu/FAQ/FAQformat.html#format4</a>) file.
The input GFF/GTF file can be plain text file, compressed
file with extension of .gz, .Z, .z, .bz, .bz2 and .bzip2,
or even a URL pointing to accessible remote files
(<a class="reference external" href="http://">http://</a>, <a class="reference external" href="https://">https://</a> and <a class="reference external" href="ftp://">ftp://</a>). Compressed remote files
are not supported.</p>
</dd>
<dt>out_gff            Output GFF/GTF file. if argument is missing, CrossMap</dt><dd><p>will write GFF/GTF file to the STDOUT.</p>
</dd>
</dl>
</dd>
<dt>options:</dt><dd><dl class="option-list">
<dt><kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></dt>
<dd><p>show this help message and exit</p>
</dd>
</dl>
<dl class="simple">
<dt>–chromid {a,s,l}  The style of chromosome IDs. “a” = “as-is”; “l” = “long</dt><dd><p>style” (eg. “chr1”, “chrX”); “s” = “short style” (eg.
“1”, “X”).</p>
</dd>
</dl>
</dd>
</dl>
</div></blockquote>
<p>Example</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ CrossMap gff  GRCh37_to_GRCh38.chain.gz Test7.hg19.gtf output.hg38
2024-01-12 09:43:53 [INFO]  Read the chain file &quot;GRCh37_to_GRCh38.chain.gz&quot;
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ol class="arabic simple">
<li><p>Each feature  (exon, intron, UTR, etc) is processed separately and independently, and we do NOT check if features originally belonging to the same gene were converted into the same gene.</p></li>
<li><p>If a user wants to lift over gene annotation files, use BED12 format.</p></li>
<li><p>If no output file was specified, the output will be printed to screen (console). In this case, items that failed to convert are also printed out.</p></li>
</ol>
</div>
</section>
<section id="convert-vcf-format-files">
<h2>Convert VCF format files<a class="headerlink" href="#convert-vcf-format-files" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41">VCF</a>
(variant call format) is a flexible and extendable line-oriented text format developed by
the <a class="reference external" href="http://www.1000genomes.org/">1000 Genome Project</a>. It is useful for representing single
nucleotide variants, indels, copy number variants, and structural variants. Chromosomes,
coordinates, and reference alleles are updated to a new assembly, and all the other fields
are not changed.</p>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">vcf</span> <span class="pre">-h</span></code> will print help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap vcf -h

usage: CrossMap vcf [-h] [--chromid {a,s,l}] [--no-comp-alleles] [--compress]
                    input.chain input.vcf refgenome.fa out_vcf

positional arguments:
  input.chain        Chain file
                     (https://genome.ucsc.edu/goldenPath/help/chain.html)
                     describes pairwise alignments between two genomes. The
                     input chain file can be a plain text file or compressed
                     (.gz, .Z, .z, .bz, .bz2, .bzip2) file.
  input.vcf          Input VCF (variant call format,
                     https://samtools.github.io/hts-specs/VCFv4.2.pdf). The
                     VCF file can be plain text file, compressed file with
                     extension of .gz, .Z, .z, .bz, .bz2 and .bzip2, or even a
                     URL pointing to accessible remote files (http://,
                     https:// and ftp://). Compressed remote files are not
                     supported.
  refgenome.fa       Chromosome sequences of target assembly in FASTA
                     (https://en.wikipedia.org/wiki/FASTA_format) format.
  out_vcf            Output VCF file.

options:
  -h, --help         show this help message and exit
  --chromid {a,s,l}  The style of chromosome IDs. &quot;a&quot; = &quot;as-is&quot;; &quot;l&quot; = &quot;long
                     style&quot; (eg. &quot;chr1&quot;, &quot;chrX&quot;); &quot;s&quot; = &quot;short style&quot; (eg.
                     &quot;1&quot;, &quot;X&quot;).
  --no-comp-alleles  If set, CrossMap does NOT check if the reference allele
                     is different from the alternate allele.
  --compress         If set, compress the output VCF file by calling the
                     system &quot;gzip&quot;.
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ol class="arabic simple">
<li><p>Genome coordinates and reference alleles will be updated to target assembly.</p></li>
<li><p>The reference genome is the genome sequences of target assembly.</p></li>
<li><p>If the reference genome sequence file (../database/genome/hg18.fa) was not indexed, CrossMap will automatically index it (only the first time you run CrossMap).</p></li>
<li><p>Output files: <em>output_file</em> and <em>output_file.unmap</em>.</p></li>
<li><p>In the output VCF file, whether the chromosome IDs contain “chr” or not depends on the format of the input VCF file.</p></li>
</ol>
</div>
<p><strong>Interpretation of Failed tags</strong>:</p>
<ul class="simple">
<li><p>Fail(Multiple_hits) : This genomic location was mapped to two or more locations to the target assembly.</p></li>
<li><p>Fail(REF==ALT) : After liftover, the reference allele is the same as the alternative allele (i.e. this is NOT an SNP/variant after liftover). In version 0.5.2, this checking can be turned off by setting ‘–no-comp-alleles’.</p></li>
<li><p>Fail(Unmap) : Unable to map this location to the target assembly.</p></li>
<li><p>Fail(KeyError) : Unable to find the contig ID (or chromosome ID) from the reference genome sequence (of the target assembly).</p></li>
</ul>
</section>
<section id="convert-maf-format-files">
<h2>Convert MAF format files<a class="headerlink" href="#convert-maf-format-files" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://docs.gdc.cancer.gov/Encyclopedia/pages/Mutation_Annotation_Format/">MAF</a>
(mutation annotation format) files are tab-delimited
files that contain somatic and/or germline mutation annotations. Please do not confuse
with the <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format5">Multiple Alignment Format</a>.</p>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">maf</span> <span class="pre">-h</span></code> will print help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap  maf -h

usage: CrossMap maf [-h] [--chromid {a,s,l}]
                    input.chain input.maf refgenome.fa build_name out_maf

positional arguments:
  input.chain        Chain file
                     (https://genome.ucsc.edu/goldenPath/help/chain.html)
                     describes pairwise alignments between two genomes. The
                     input chain file can be a plain text file or compressed
                     (.gz, .Z, .z, .bz, .bz2, .bzip2) file.
  input.maf          Input MAF (https://docs.gdc.cancer.gov/Data/File_Formats/
                     MAF_Format/) format file. The MAF file can be plain text
                     file, compressed file with extension of .gz, .Z, .z, .bz,
                     .bz2 and .bzip2, or even a URL pointing to accessible
                     remote files (http://, https:// and ftp://). Compressed
                     remote files are not supported.
  refgenome.fa       Chromosome sequences of target assembly in FASTA
                     (https://en.wikipedia.org/wiki/FASTA_format) format.
  build_name         the name of the *target_assembly* (eg &quot;GRCh38&quot;).
  out_maf            Output MAF file.

options:
  -h, --help         show this help message and exit
  --chromid {a,s,l}  The style of chromosome IDs. &quot;a&quot; = &quot;as-is&quot;; &quot;l&quot; = &quot;long
                     style&quot; (eg. &quot;chr1&quot;, &quot;chrX&quot;); &quot;s&quot; = &quot;short style&quot; (eg.
                     &quot;1&quot;, &quot;X&quot;).
</pre></div>
</div>
</section>
<section id="convert-gvcf-format-files">
<h2>Convert GVCF format files<a class="headerlink" href="#convert-gvcf-format-files" title="Permalink to this heading"></a></h2>
<p>GVCF file format is described in <a class="reference external" href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531812-GVCF-Genomic-Variant-Call-Format">here</a>.</p>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">gvcf</span> <span class="pre">-h</span></code> will print help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap  gvcf -h

usage: CrossMap gvcf [-h] [--chromid {a,s,l}] [--no-comp-alleles] [--compress]
                     input.chain input.gvcf refgenome.fa out_gvcf

positional arguments:
  input.chain        Chain file
                     (https://genome.ucsc.edu/goldenPath/help/chain.html)
                     describes pairwise alignments between two genomes. The
                     input chain file can be a plain text file or compressed
                     (.gz, .Z, .z, .bz, .bz2, .bzip2) file.
  input.gvcf         Input gVCF (genomic variant call format,
                     https://samtools.github.io/hts-specs/VCFv4.2.pdf). The
                     gVCF file can be plain text file, compressed file with
                     extension of .gz, .Z, .z, .bz, .bz2 and .bzip2, or even a
                     URL pointing to accessible remote files (http://,
                     https:// and ftp://). Compressed remote files are not
                     supported.
  refgenome.fa       Chromosome sequences of target assembly in FASTA
                     (https://en.wikipedia.org/wiki/FASTA_format) format.
  out_gvcf           Output gVCF file.

options:
  -h, --help         show this help message and exit
  --chromid {a,s,l}  The style of chromosome IDs. &quot;a&quot; = &quot;as-is&quot;; &quot;l&quot; = &quot;long
                     style&quot; (eg. &quot;chr1&quot;, &quot;chrX&quot;); &quot;s&quot; = &quot;short style&quot; (eg.
                     &quot;1&quot;, &quot;X&quot;).
  --no-comp-alleles  If set, CrossMap does NOT check if the reference allele
                     is different from the alternate allele.
  --compress         If set, compress the output VCF file by calling the
                     system &quot;gzip&quot;.
</pre></div>
</div>
</section>
<section id="convert-large-genomic-regions">
<h2>Convert large genomic regions<a class="headerlink" href="#convert-large-genomic-regions" title="Permalink to this heading"></a></h2>
<p>For <strong>large genomic regions</strong> such as CNV blocks, the <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">bed</span></code> will split each large region into smaller blocks that are 100% matched to the target assembly.
<code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">region</span></code> will NOT split large regions, instead, it will calculate the <strong>map ratio</strong> (i.e. {bases mapped to target genome} / {total bases in query region}). If the
<strong>map ratio</strong> is larger than the threshold specified by <code class="code docutils literal notranslate"><span class="pre">-r</span></code>, the coordinates will be converted to the target genome, otherwise, it fails.</p>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">region</span> <span class="pre">-h</span></code> will print help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>usage: CrossMap region [-h] [--chromid {a,s,l}] [-r MIN_MAP_RATIO]
                       input.chain input.bed [out_bed]

positional arguments:
  input.chain           Chain file
                        (https://genome.ucsc.edu/goldenPath/help/chain.html)
                        describes pairwise alignments between two genomes. The
                        input chain file can be a plain text file or
                        compressed (.gz, .Z, .z, .bz, .bz2, .bzip2) file.
  input.bed             The input BED file. The first 3 columns must be
                        “chrom”, “start”, and “end”. The input BED file can be
                        plain text file, compressed file with extension of
                        .gz, .Z, .z, .bz, .bz2 and .bzip2, or even a URL
                        pointing to accessible remote files (http://, https://
                        and ftp://). Compressed remote files are not
                        supported.
  out_bed               Output BED file. if argument is missing, CrossMap will
                        write BED file to the STDOUT.

options:
  -h, --help            show this help message and exit
  --chromid {a,s,l}     The style of chromosome IDs. &quot;a&quot; = &quot;as-is&quot;; &quot;l&quot; =
                        &quot;long style&quot; (eg. &quot;chr1&quot;, &quot;chrX&quot;); &quot;s&quot; = &quot;short style&quot;
                        (eg. &quot;1&quot;, &quot;X&quot;).
  -r MIN_MAP_RATIO, --ratio MIN_MAP_RATIO
                        Minimum ratio of bases that must remap.
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ol class="arabic simple">
<li><p>Input BED file should have at least 3 columns (chrom, start, end). Additional columns will be kept as is.</p></li>
</ol>
</div>
</section>
<section id="view-chain-file">
<h2>View chain file<a class="headerlink" href="#view-chain-file" title="Permalink to this heading"></a></h2>
<p>Typing <code class="code docutils literal notranslate"><span class="pre">CrossMap</span> <span class="pre">viewchain</span> <span class="pre">-h</span></code> will print help message:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">CrossMap</span> <span class="n">viewchain</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="nb">input</span><span class="o">.</span><span class="n">chain</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="nb">input</span><span class="o">.</span><span class="n">chain</span>  <span class="n">Chain</span> <span class="n">file</span> <span class="p">(</span><span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">genome</span><span class="o">.</span><span class="n">ucsc</span><span class="o">.</span><span class="n">edu</span><span class="o">/</span><span class="n">goldenPath</span><span class="o">/</span><span class="n">help</span><span class="o">/</span><span class="n">chain</span><span class="o">.</span><span class="n">html</span><span class="p">)</span>
               <span class="n">describes</span> <span class="n">pairwise</span> <span class="n">alignments</span> <span class="n">between</span> <span class="n">two</span> <span class="n">genomes</span><span class="o">.</span> <span class="n">The</span> <span class="nb">input</span>
               <span class="n">chain</span> <span class="n">file</span> <span class="n">can</span> <span class="n">be</span> <span class="n">a</span> <span class="n">plain</span> <span class="n">text</span> <span class="n">file</span> <span class="ow">or</span> <span class="n">compressed</span> <span class="p">(</span><span class="o">.</span><span class="n">gz</span><span class="p">,</span> <span class="o">.</span><span class="n">Z</span><span class="p">,</span> <span class="o">.</span><span class="n">z</span><span class="p">,</span>
               <span class="o">.</span><span class="n">bz</span><span class="p">,</span> <span class="o">.</span><span class="n">bz2</span><span class="p">,</span> <span class="o">.</span><span class="n">bzip2</span><span class="p">)</span> <span class="n">file</span><span class="o">.</span>

<span class="n">options</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>   <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>
</pre></div>
</div>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ CrossMap viewchain GRCh37_to_GRCh38.chain.gz &gt;chain.tab
$ head chain.tab
1  10000 177417   +  1  10000 177417   +
1  227417   267719   +  1  257666   297968   +
1  317719   471368   +  1  347968   501617   -
1  521368   1566075  +  1  585988   1630695  +
1  1566075  1569784  +  1  1630696  1634405  +
1  1569784  1570918  +  1  1634408  1635542  +
1  1570918  1570922  +  1  1635546  1635550  +
1  1570922  1574299  +  1  1635560  1638937  +
1  1574299  1583669  +  1  1638938  1648308  +
1  1583669  1583878  +  1  1648309  1648518  +
</pre></div>
</div>
</section>
</section>
<section id="compare-to-ucsc-liftover-tool">
<h1>Compare to UCSC liftover tool<a class="headerlink" href="#compare-to-ucsc-liftover-tool" title="Permalink to this heading"></a></h1>
<p>To assess the accuracy of CrossMap, we randomly generated 10,000 genome intervals (download from <a class="reference external" href="https://sourceforge.net/projects/crossmap/files/hg19.rand.bed.gz/download">here</a>) with the
fixed interval size of 200 bp from hg19. Then we converted them into hg18 using CrossMap
and <a class="reference external" href="http://genome.ucsc.edu/cgi-bin/hgLiftOver">UCSC liftover tool</a> with default configurations. We compare CrossMap
to <a class="reference external" href="http://genome.ucsc.edu/cgi-bin/hgLiftOver">UCSC liftover tool</a> because it is the most widely
used tool to convert genome coordinates.</p>
<p>CrossMap failed to convert 613 intervals, and the UCSC liftover tool failed to convert 614
intervals. All failed intervals are exactly the same except for one region (chr2 90542908 90543108).
UCSC failed to convert it because this region needs to be split twice:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Original (hg19)</p></th>
<th class="head"><p>Split (hg19)</p></th>
<th class="head"><p>Target (hg18)</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>chr2 90542908  90543108 -</p></td>
<td><p>chr2 90542908 90542933 -</p></td>
<td><p>chr2    89906445        89906470 -</p></td>
</tr>
<tr class="row-odd"><td><p>chr2 90542908  90543108 -</p></td>
<td><p>chr2 90542933 90543001 -</p></td>
<td><p>chr2    87414583        87414651 -</p></td>
</tr>
<tr class="row-even"><td><p>chr2 90542908  90543108 -</p></td>
<td><p>chr2 90543010 90543108 -</p></td>
<td><p>chr2    87414276        87414374 -</p></td>
</tr>
</tbody>
</table>
<p>For genome intervals that were successfully converted to hg18, the start and end coordinates are
exactly the same between UCSC conversion and CrossMap conversion.</p>
<a class="reference internal image-reference" href="_images/CrossMap_vs_UCSC.png"><img alt="CrossMap_vs_UCSC_liftover.png" src="_images/CrossMap_vs_UCSC.png" style="width: 700.0px; height: 400.0px;" /></a>
</section>
<section id="citation">
<h1>Citation<a class="headerlink" href="#citation" title="Permalink to this heading"></a></h1>
<p><a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/24351709/">Zhao, H., Sun, Z., Wang, J., Huang, H., Kocher, J.-P., &amp; Wang, L. (2013). CrossMap: a versatile tool for coordinate conversion between genome assemblies. Bioinformatics (Oxford, England), btt730</a></p>
</section>
<section id="license">
<h1>LICENSE<a class="headerlink" href="#license" title="Permalink to this heading"></a></h1>
<p>CrossMap is distributed under <a class="reference external" href="http://www.gnu.org/copyleft/gpl.html">GNU General Public License</a></p>
<p>This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version. This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details. You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301 USA</p>
</section>
<section id="contact">
<h1>Contact<a class="headerlink" href="#contact" title="Permalink to this heading"></a></h1>
<ul class="simple">
<li><p>Wang.Liguo AT mayo.edu</p></li>
</ul>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Liguo Wang.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-XXXXXXX-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'UA-XXXXXXX-1', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>